cmd/pinniped-proxy/src/pinniped.rs

```diff
@@ -0,0 +1,141 @@
+/// resources based on the struct. See https://docs.rs/kube/0.43.0/kube/ for more details.
+#[derive(CustomResource, Deserialize, Serialize, Clone, Debug)]
+#[kube(group = "login.concierge.pinniped.dev", version = "v1alpha1", kind = "TokenCredentialRequest", namespaced)]
+#[kube(status = "TokenCredentialRequestStatus")]
+pub struct TokenCredentialRequestSpec {
+    // Bearer token supplied with the credential request.
+    token: Option<String>,
+
+    // Reference to an authenticator which can verify this credential request.
+    authenticator: corev1::TypedLocalObjectReference,
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+pub struct TokenCredentialRequestStatus {
+    // A ClusterCredential will be returned for a successful credential request.
+    credential: Option<ClusterCredential>,
+
+    // An error message will be returned for an unsuccessful credential request.
+    message: Option<String>,
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+/// ClusterCredential is the cluster-specific credential returned on a successful credential request. It
+/// contains either a valid bearer token or a valid TLS certificate and corresponding private key for the cluster.
+#[serde(rename_all = "camelCase")]
+pub struct ClusterCredential {
+    // ExpirationTimestamp indicates a time when the provided credentials expire.
+    expiration_timestamp: metav1::Time,
+
+    // Token is a bearer token used by the client for request authentication.
+    token: Option<String>,
+
+    // PEM-encoded client TLS certificates (including intermediates, if any).
+    client_certificate_data: String,
+
+    // PEM-encoded private key for the above certificate.
+    client_key_data: String,
+}
+
+/// call_pinniped_exchange returns the resulting TokenCredentialRequest with Status after requesting a token credential exchange.
+async fn call_pinniped_exchange(authorization: &str, k8s_api_server_url: &str, k8s_api_ca_cert_data: &[u8]) -> Result<TokenCredentialRequest> {
+    let pinniped_namespace = env::var("PINNIPED_NAMESPACE")?;
+
+    let mut config = Config::new(Url::parse(k8s_api_server_url).context("Failed parsing url for exchange")?);
+    config.default_ns = pinniped_namespace.clone();
+    let cert = reqwest::Certificate::from_pem(k8s_api_ca_cert_data)?;
+    config.root_cert = Some(vec![cert]);
+    let client = Client::new(config);
+
+    let auth_token = match authorization.to_string().strip_prefix("Bearer ") {
+        Some(a) => a.to_string(),
+        None => authorization.to_string(),
+    };
+    let token_creds: Api<TokenCredentialRequest> = Api::namespaced(client.clone(), &pinniped_namespace);
+    let mut cred_request = TokenCredentialRequest::new("", TokenCredentialRequestSpec {
+        token: Some(auth_token),
+        authenticator: corev1::TypedLocalObjectReference {
+            name: env::var(PINNIPED_AUTHENTICATOR_NAME).with_context(|| format!("error retrieving {}", PINNIPED_AUTHENTICATOR_NAME))?,
```

Comment:
`with_context` is a helper of the anyhow error library that enables changed errors, a little like golang's newer errors. So for example, if I run with the incorrect authenticator type var, I'll see:

```
 ERROR pinniped_proxy::service > Failed to exchange credentials

Caused by:
    0: error retrieving PINNIPED_AUTHENTICATOR_TYPE
    1: environment variable not found
```